appinventor/components/src/com/google/appinventor/components/runtime/Form.java

```diff
@@ -695,8 +696,25 @@ void ReplayFormOrientation() {
     }
   }
 
+  private Integer generateHashCode(AndroidViewComponent component, PercentStorageRecord.Dim dim) {
+    if (dim == PercentStorageRecord.Dim.HEIGHT) {
+      return component.hashCode() * 2 + 1;
+    } else {
+      return component.hashCode() * 2;
+    }
+  }
+
   public void registerPercentLength(AndroidViewComponent component, int length, PercentStorageRecord.Dim dim) {
-    dimChanges.add(new PercentStorageRecord(component, length, dim));
+    PercentStorageRecord r = new PercentStorageRecord(component, length, dim);
+    Integer key = generateHashCode(component, dim);
+    dimChanges.put(key, r);
+  }
+
+  public void unregisterPercentLength(AndroidViewComponent component, PercentStorageRecord.Dim dim) {
+    // iterate map, remove all entry match this
+    for (Integer key : dimChanges.keySet()) {
```

Comment:
You don't need the for loop here any more. The call to `remove` on line 716 either will remove the entry if it exists or do nothing if it doesn't exist. Because it changes the map, it will result in a `ConcurrentModificationException` as written since you're modifying the loop while iterating over it.